src/js/facility-locator/components/Select.jsx

```diff
@@ -0,0 +1,161 @@
+import React, { Component } from 'react';
+import classNames from 'classnames';
+import Listbox from './Listbox';
+import {
+  getDirection,
+  getSelection,
+  isSelect,
+  isToggle,
+  isTraverse,
+  noServices
+} from '../utils/helpers.js';
+
+class Select extends Component {
+  constructor() {
+    super();
+
+    this.state = {
+      dropdownActive: false,
+      focusedIndex: 0
+    };
+    this.options = [];
+  }
+
+  handleFilterSelect = (newOption, type) => {
+    const { currentQuery } = this.props;
+    const queryType = currentQuery[`${type}Type`];
+    if (type === 'facility' && newOption !== queryType) {
+      if (newOption === 'AllFacilities') {
+        this.props.updateSearchQuery({
+          facilityType: null,
+          serviceType: null
+        });
+      } else {
+        this.props.updateSearchQuery({
+          facilityType: newOption,
+          serviceType: null
+        });
+      }
+    }
+    if (type === 'service' && newOption !== queryType) {
+      if (newOption === 'All') {
+        this.props.updateSearchQuery({
+          serviceType: null,
+        });
+      } else {
+        this.props.updateSearchQuery({
+          serviceType: newOption
+        });
+      }
+    }
+    return this.toggleDropdown(null, type);
+  }
+
+  navigateDropdown = (e) => {
+    const type = this.props.optionType;
+    const isActive = this.state.dropdownActive;
+    const which = e.target;
+    if (isToggle(e, isActive)) {
+      if (isActive && isSelect(e.keyCode)) {
+        this.handleFilterSelect(which.id, type);
+      }
+      return this.toggleDropdown(e, type);
+    }
+    if (isActive && isTraverse(e)) {
+      const increment = getDirection(e.keyCode);
+      const newIndex = this.state.focusedIndex + increment;
+      return this.focusOption(this.options[newIndex], newIndex, type);
+    }
+    return false;
+  }
+
+  handleBlur = (e) => {
+    const { currentTarget } = e;
+    setTimeout(() => {
+      if (!currentTarget.contains(document.activeElement)) {
+        this.setState({ dropdownActive: false });
+      }
+    });
+  }
+
+  toggleDropdown = (e) => {
+    const type = this.props.optionType;
+    const isActive = this.state.dropdownActive;
+    const queryType = this.props.currentQuery[`${type}Type`];
+    // To handle unrelated key events
+    if (e) {
+      const shouldNotToggle = e.keyCode && !isToggle(e, isActive);
+      if (shouldNotToggle) return;
+    }
+    if (isActive) {
+      this.dropdown.focus();
+    } else {
+      const { selection, id } = getSelection(this.options, queryType);
+      this.focusOption(selection, id, type);
+    }
+    this.setState({
+      dropdownActive: !this.state.dropdownActive
+    });
+  }
+
+  focusOption = (option, index) => {
+    if (option) {
+      option.focus();
+    }
+    if (this.options[index]) {
```

Comment:
Can you have a case where this method is called with a valid option to focus on, but it's not in the options array? It's not clear to me why there are two slightly different conditionals in here.

Reply:
You are right, this is paranoid programming.